### ISSUE TITLE
Porting to .net 

### ISSUE THREAD

**Issue by Danielku15** [schellingb/TinySoundFont#32]
Hi there. 

**tl;dr:** Would you be fine for you, if I port your library to C# and merge it into my music notation library I am developing? I will try to retain the original copyright notice as good as possible but the borders between my library and yours will likely become quite fuzzy over time. 

**Long Version:** 

First of all thanks for this great library. Most of the synthesizer libraries out there are really heavyweight and consist of thousands of lines of code but your library is really slim and produces great results. The best and most complete ones out there, are still FluidSynth and Gervill (part of the JDK) but with simple soundfonts you can hardly here a difference (beside the missing effects #24 ). FluidSynth is quite heavyweight and hard to port to other languages due to its C-nature, huge codebase and tons of capabilities. Gervill with it's GPL license and Oracle in the back, I can already forget to use it outside of the Java world. 

I am the author of [alphaTab](https://github.com/CoderLine/alphaTab/), a web (and .net) based music notation software with playback capabilities. I already have a soundfont based synthesis library where I generate raw audio samples based on a given midi input to play it via the audio interface available on the platform. TinySoundFont has a way smaller footprint and produces more accurate results than my current one and I'm considering to replace my synth library partly with yours.  

In general the MIT license would not prevent me to port your library to another language, but I prefer to talk to the original authors of libraries before doing so to get official permission. Especially as it will not remain a simple 1:1 port of the original library. 

My plan is to port the library to C# and then further cross compile it to JavaScript and Kotlin using a home-made compiler/transpiler. Actually I already finished the porting and it runs already on .net 😉. Unfortunately I will not be able to use the library as-it-is and I will need to make further extensions for my API and the need of SoundFont 3 (Ogg Vorbis) support. So I will need to rip apart the structure a bit and introduce new functionality here and there. Also my library currently is released under LGPL which would be then more restrictive. 

Of course I will add to all synthesizer related classes your name, and refer to your project as well to maintain the original copyright notice but somehow TinySoundFont will likely be hidden within my library. 

I'm also looking forward to implement certain missing parts of the synthesizer on top of my port. I compared several midi file/soundfont combinations between TinySoundFont and Gervill and often they sound quite different. I hope to eliminate those differences. 

I'm looking forward to get your permission to adopt your library and integrate it with mine. 

Kind Regards
Daniel

**Comment by Danielku15** [comment]
Thanks for the info. The attribution would be explicitely in the source files that are based on this project + in the source header of the final javascript. The .net redistribution happens via the NuGet package manager which again refers to the original license description. Of course, people who finally will use my library, are responsible again to ensure a correct attribution of my library which then must include the correct copyright notice which will again include all attributions. But as you say: Correct attribution rarely happens in the market especially in the web with all minimizers. 

I will try to keep you informed on bigger changes I make and could be ported back to the C-version of the source. Especially things related to the actual audio generation (missing handling of modulator, generators or zone handling) should be easily portable as the core of the audiogeneration will stay. 

For SF3 support you basically need to decode Ogg Vorbis files. The sample headers have then a special flag set and you need to decode a certain buffer range of the original sample buffer and adapt some start/end loopstart/loopend values. The most tricky thing I am doing, is to decode only the samples of the presets that are used in the midi file.

**Comment by Danielku15** [comment]
Hello again. 
It's been quite a while since I requested for permission on using TSF in AlphaTab. After I did tons of investigations during the last months on various synthesis engines I finally integrated TSF into AlphaTab. 

The related PR on my side is https://github.com/CoderLine/alphaTab/pull/261 
I tried to attribute your project (and SFZero) as good as possible and reasonable. 

- The source files that were ported from TSF contain a dedicated file header annotating the original copyright. 
- The compile JavaScript file also contains a dedicated attributed. 
- On the project readme I added an own point to the "Thanks to" section. 

I yesterday changed my project license to MPL-2.0 which is less restrictive and better compatible with MIT too. 

I would like to get your final confirmation if the .net port of the library is okay for you like that. 

I did not make any bigger extensions or modifications to the library beside: 
- Adding a feature where I can mute/solo individual channels. 
- Set a separate use selected mix-volume per channel 

My investigations the last months have shown that the biggest missing part to get the TSF synthesized audio sound correctly are the modulators. I am still learning how the modulators actually are modifying the raw samples before mixing and how to get this into the TSF port. This will become the first biggest extension to TSF on my side. I will try to port back my changes to your library if I am successful.

**Comment by schellingb** [comment]
Hey there again!
Congratulations on finishing it and getting it working in C#!
Thank you for the attribution and again for the thoughts put into it. I'm perfectly okay with it. Maybe some day @stevefolta will find this topic and can give feedback as well :-)

Modulators are certainly the biggest missing feature for full SoundFont 2.01 compatibility. The few times I pondered implementation it seemed to me that it would need more code than TinySoundFont itself to support it and achieving good compatibility with the specs seems rather difficult as well. Meanwhile I haven't seen many soundfonts use them at all. Maybe it's more common for very high quality fonts?
But if you have a sweet implementation, I'd gladly consider adding it to TSF.

Personally I'm more interested in a good and small chorus and reverb filter that would fit in nicely with TSF.

**Comment by Danielku15** [comment]
Thanks for the fast feedback. I will try to ping Steve too, but he seems not to be very active on GitHub anymore and his website is also not updated at all. I doubt he will call back. 

I mostly noticed the missing modulators on high quality SoundFonts but also on some smaller ones they were used to make slight adjustments to volumes. The lack of modulators made some instruments sometimes too loud compared to the expected playback. Especially also considering there is quite a set of default modulators that should be applied. I will try to keep you updated on that topic. But firstly I want to get my next library release finalized before I start new topics like this 😉
